### PR TITLE
Initial commit of the awscloudwatchlogsexporter

### PR DESCRIPTION
**Pull request by rakyll**
This change adds an exporter to export OpenTelemetry logs to CloudWatch Logs.

Given that OpenTelemetry Logging support is subject to change, this package is not currently stable and might break APIs or behavior in the future.

### CODE REVIEW

**Review by anuraaga — COMMENTED** [review]
Cool

File: exporter/awscloudwatchlogsexporter/exporter.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+		SeverityNumber:         int32(log.SeverityNumber()),
+		SeverityText:           log.SeverityText(),
+		DroppedAttributesCount: uint32(log.DroppedAttributesCount()),
+		Flags:                  log.Flags(),
+	}
+	if traceID := log.TraceID(); traceID.IsValid() {
+		body.TraceID = traceID.HexString()
+	}
+	if spanID := log.SpanID(); spanID.IsValid() {
+		body.TraceID = spanID.HexString()
```
Comment (anuraaga) [inline exporter/awscloudwatchlogsexporter/exporter.go:190]:
```suggestion
		body.SpanID = spanID.HexString()
```

File: exporter/awscloudwatchlogsexporter/exporter.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	if log.Attributes().Len() > 0 {
+		attrs := make(map[string]interface{}, log.Attributes().Len())
+		log.Attributes().ForEach(func(k string, v pdata.AttributeValue) {
+			attrs[k] = attrValue(v)
+		})
+		body.Attributes = attrs
+	}
+
+	// Add resource attributes.
+	if resource.Attributes().Len() > 0 {
```
Comment (anuraaga) [inline exporter/awscloudwatchlogsexporter/exporter.go:202]: We may as well do this conversion outside this method so it's shared among all the resourcespans

File: exporter/awscloudwatchlogsexporter/exporter.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+		})
+		body.Resource = attrs
+	}
+
+	bodyJSON, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+	return &cloudwatchlogs.InputLogEvent{
+		Timestamp: aws.Int64(int64(log.Timestamp()) / (1000 * 1000)), // milliseconds
```
Comment (anuraaga) [inline exporter/awscloudwatchlogsexporter/exporter.go:215]: Is it possible to use `time.Millisecond`?

File: exporter/awscloudwatchlogsexporter/exporter.go (hunk 1 above)
Comment (anuraaga) [reply exporter/awscloudwatchlogsexporter/exporter.go:190]: Possibly missing test coverage

**Comment by rakyll** [comment]
PTAL

I can't see the build logs so don't know why it's failing.

**Comment by anuraaga** [comment]
@rakyll I don't see any changes for my most recent comments, perhaps something wrong when pushing?

Also I figured you'd be able to see the circleci logs after logging in with GitHub. But just incase, something wrong with the Makefile I think

```
#!/bin/bash -eo pipefail
make -j8 for-all-target TARGET=dep
make -C ./exporter/alibabacloudlogserviceexporter dep
make -C ./exporter/awscloudwatchlogsexporter dep
make -C ./exporter/awsemfexporter dep
make -C ./exporter/awsprometheusremotewriteexporter dep
make -C ./exporter/awsxrayexporter dep
make -C ./exporter/azuremonitorexporter dep
make[1]: Entering directory '/home/circleci/project/exporter/alibabacloudlogserviceexporter'
make -C ./exporter/carbonexporter dep
make -C ./exporter/datadogexporter dep
make[1]: Entering directory '/home/circleci/project/exporter/awsemfexporter'
make[1]: Entering directory '/home/circleci/project/exporter/awsxrayexporter'
make[1]: Entering directory '/home/circleci/project/exporter/awsprometheusremotewriteexporter'
make[1]: Entering directory '/home/circleci/project/exporter/awscloudwatchlogsexporter'
make[1]: *** No rule to make target 'dep'.  Stop.
make[1]: Leaving directory '/home/circleci/project/exporter/awscloudwatchlogsexporter'
make[1]: Entering directory '/home/circleci/project/exporter/azuremonitorexporter'
make[1]: Entering directory '/home/circleci/project/exporter/carbonexporter'
make: *** [Makefile:111: for-all-target-./exporter/awscloudwatchlogsexporter] Error 2
make: *** Waiting for unfinished jobs....
make[1]: Entering directory '/home/circleci/project/exporter/datadogexporter'
go mod download
go mod download
go mod download
go mod download
go mod download
go mod download
go mod download
make[1]: Leaving directory '/home/circleci/project/exporter/awsprometheusremotewriteexporter'
make[1]: Leaving directory '/home/circleci/project/exporter/azuremonitorexporter'
make[1]: Leaving directory '/home/circleci/project/exporter/carbonexporter'
make[1]: Leaving directory '/home/circleci/project/exporter/awsemfexporter'
make[1]: Leaving directory '/home/circleci/project/exporter/alibabacloudlogserviceexporter'
make[1]: Leaving directory '/home/circleci/project/exporter/awsxrayexporter'
make[1]: Leaving directory '/home/circleci/project/exporter/datadogexporter'

Exited with code exit status 2
CircleCI received exit code 2
```

File: exporter/awscloudwatchlogsexporter/exporter.go (hunk 3 above)
Comment (rakyll) [reply exporter/awscloudwatchlogsexporter/exporter.go:215]: Yo mean, int64(log.Timestamp()) / int64(time.Millisecond)? 

I don't think it's very readable.

File: exporter/awscloudwatchlogsexporter/exporter.go (hunk 2 above)
Comment (rakyll) [reply exporter/awscloudwatchlogsexporter/exporter.go:202]: Done

File: exporter/awscloudwatchlogsexporter/exporter.go (hunk 1 above)
Comment (rakyll) [reply exporter/awscloudwatchlogsexporter/exporter.go:190]: Fixed, the test was broken too.

**Comment by rakyll** [comment]
Addressed the comments and looking at the broken step now.

File: exporter/awscloudwatchlogsexporter/exporter.go (hunk 3 above)
Comment (anuraaga) [reply exporter/awscloudwatchlogsexporter/exporter.go:197]: Well it's self-documenting and compile-time safe. We have a comment for `// milliseconds` anyways so the current code seems to acknowledge it's a bit tricky.

File: exporter/awscloudwatchlogsexporter/exporter.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	if n == 0 {
+		return []*cloudwatchlogs.InputLogEvent{}, 0
+	}
+
+	var dropped int
+	out := make([]*cloudwatchlogs.InputLogEvent, 0) // TODO(jbd): set a better capacity
+
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		rl := rls.At(i)
```
Comment (anuraaga) [inline exporter/awscloudwatchlogsexporter/exporter.go:139]: This comment might have been confusing

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1944#discussion_r552299536

I meant we should convert `rl.Resource()` into `map` here and pass the map into `logToCWLog`. The resource is constant, no need to convert for each log in the same resource.

**Comment by rakyll** [comment]
@anuraaga Do you think we should merge something like this as an experiment? I'm currently using it to dogfood the logs data model.

**Comment by anuraaga** [comment]
@tigrannajaryan @bogdandrutu I personally would be fine with merging but want to make sure it doesn't conflict with the stability plan of the collector. I guess it's only about traces so adding a logs exporter should be ok.

**Comment by tigrannajaryan** [comment]
Since this is a separate component in contrib, it does not interfere with the stability plan. I haven't reviewed this PR, but I don't see a reason why we would want to block such PRs. Feel free to reopen.

**Comment by rakyll** [comment]
@anuraaga @tigrannajaryan We need to consolidate on a serialization format before we merge this. I don't want to make this available and change the output format from JSON and break the users. Give us some time, I'll open it again once we agree on the format.

**Comment by hossain-rayhan** [comment]
I would also love to understand why do we need two different exporters for sending telemetry data to CloudWatch. We have the `awsemf` exporter which sends metrics to CloudWatch. I believe we are calling the same API for sending logs as well and we have seperate interfaces for handling logs and metrics in the exporter architecture. Also, EMF format is CloudWatch logs and different in terms of implementation detail.

**Pull request closed without merging**